### DISCUSSION TITLE
useArrayField nested with separate files

### DISCUSSION THREAD

**Discussion by GabrielGomesBarbosa** [react-hook-form/react-hook-form#2283] [Q&A] [answered]
Hi @bluebill1049 

I have a question here, about **useFieldArray** with components, i don't think that is necessary put the code in sandbox, but if you want, i can do this... but i trying to use with components in different files and dont have it on documentation.

Example: 

Component Employee.jsx

```
const Employee = props => {

     const { setValue, watch } = useFormContext();

    const { fields, append, remove } = useFieldArray({
        name: "employee",
        keyName: "key"
     });    
    

    //Here i can map fields

     {
                fields.map((item, index) => {
                    return (
                          <>
                            <input 
                                type="hidden" 
                                name={`employee[${index}].id`} 
                                ref={register()}
                                defaultValue={`${item.id}`}
                            />  
                           <ItemHour parentIndex={index} />
                         </>
                    )
                })
            }
}
```

Now i want to create the **ItemHour** component that will have the nested fields from employee component, should a initializate the **useFieldArray** with the same key of the parent? in this case... everything is inside <FormContext>

```
const HourItem = props => {

     const { setValue, watch } = useFormContext();
     
    //This repeating
    const { fields, append, remove } = useFieldArray({
        name: "employee",
        keyName: "key"
     });    
     
     //Here i want to put more fields inside employee object like this:
     // employee: {
     //id: "", --> is in parent
     //hour: [] --> will be create here
     //} 

    useEffect(() => {
        append({startDate: "start", endDate: "end" });
    }, []);

    //Here is the same interaction 
    {
             fields.map((item, index) => {
                        return (
                            <Row key={item.key}>
                                <input 
                                    type="text" 
                                    name={`employee[${props.parentIndex}].hour.startDate`} 
                                    ref={register()}
                                    defaultValue={`${item.startDate} - ${props.parentIndex}${index}`}
                                />
                                <input 
                                    type="text" 
                                    name={`employee[${props.parentIndex}].hour.endDate`} 
                                    ref={register()}
                                    defaultValue={`${item.endDate} - ${props.parentIndex}${index}`}
                                />
                            </Row>
                        );
          }
}

```

In **HourItem** component the **useEffect** add a new line, bute the value dont stay the correct object, so i cant initialize with a value.

I have read your example in here, https://codesandbox.io/s/vy8fv but dont have nested field in other component file... im using v5.

Can you help me with this ?

I want to know:
If this is correct way to use useFormArray in the component that is a separate file to keep same name field.
Correct way to initialize with values.

**Comment by bluebill1049** [comment]
hey, there is an example in the example folder as well :) 

Array of Field Array : https://github.com/react-hook-form/react-hook-form/tree/master/examples

https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr

**Reply by GabrielGomesBarbosa** [reply]
Thanks @bluebill1049 i will study it.

**Comment by GabrielGomesBarbosa** [comment]
@bluebill1049 in this case https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr you pass the default values using **useForm()** in my case i'm using **useFormContext()** so dont have a constructor with default values. like this:

```
    const { setValue, watch } = useFormContext(defaultValues);
     
    const { fields, append, remove } = useFieldArray({
        name: "employee",
        keyName: "key"
     });    
```
how can i pass the value to nested useArrayField ?

im trying to receive default values in props and iterate it on **useEffect** like this:

```
useEffect(() => {
    props.defaultValues.forEach(item => {
        append(item) 
   })
}, []);
```

but something went wrong, because if the array from default values has lenght equals 3 just show 2 itens on screen, its always jump the first item.

```
useEffect(() => {
    props.defaultValues.forEach(item => {
        append({employee: item}) 
   })
}, []);
```
and i tryed this too.

**Reply by bluebill1049** [reply]
```
useForm({
  defaultValues
})
```
why not do it in `useForm` level? otherwise you can always using `reset` API to reset the whole Field Array.

**Comment by GabrielGomesBarbosa** [comment]
Hi @bluebill1049 i still need your help, when i pass the values inside **useForm()** its work, look at this scenario: my form is inside a modal, and that modal there is you own state with prop `isOpen` so... and inside the modal i have the constructor of **useForm** too, the ideia is pass tho the modal the props `isOpen = true` and the prop `defaultValues` look:

```
const Modal = ({ title, isOpen, children, closeEvent, submitEvent, defaultValues }) => {

    console.log("Default Values", defaultValues);
    const formMethods = useForm({
        defaultValues
    });

    return (
        isOpen && <Container>
            <Header 
                title={title} 
                close={closeEvent}
            />
            <Body>
                <FormContext {...formMethods}>
                    <form>
                        <Row>
                            {children}
                        </Row>
                    </form>
                </FormContext>
            </Body>
            <Footer 
                close={closeEvent} 
                handleSubmit={formMethods.handleSubmit(submitEvent)}
            />
        </Container>
    )
}; 
```

but when i get response from server and pass via props... the `console.log()` show all the values but the fields are not filled, but if i put some mock data inside the modal just to test tha fields are filled. Like this:

```
const Modal = ({ title, isOpen, children, closeEvent, submitEvent, defaultValues }) => {

     const defaultValues2 = {
          field1: "value",
          field2: "value",
          field3: [{test: "test"}, {name: "name"}]
    };

    const formMethods = useForm({
        defaultValues: defaultValues2
    });
```
☝️ Its work! but i need to pass via `props`

What the correct way to do that?

**Comment by bluebill1049** [comment, marked as answer]
If it's a pop-up, try to use reset at useEffect to reset those default Values.

**Reply by GabrielGomesBarbosa** [reply]
@bluebill1049 really sorry, i found it on documentation https://react-hook-form.com/pt/v5/api#reset, thanks for everything!

**Reply by bluebill1049** [reply]
all good mate 👍
